resource_customizations/numaplane.numaproj.io/ISBServiceRollout/health.lua

```diff
@@ -2,31 +2,62 @@ local hs = {}
 local healthyCondition = {}
 
 if obj.status ~= nil then
+
+  -- check for certain cases of "Progressing"
+
+  if obj.metadata.generation ~= obj.status.observedGeneration then
+    hs.status = "Progressing"
+    hs.message = "Not yet reconciled"
+    return hs
+  end
+
+  if obj.status.phase == "Pending" then
+    hs.status = "Progressing"
+    hs.message = "Phase=Pending"
+    return hs
+  end
+
+  if obj.status.upgradeInProgress ~= nil and obj.status.upgradeInProgress ~= "" then
+    hs.status = "Progressing"
+    hs.message = "Update in progress"
+    return hs
+  end
+
+  -- now check the Conditions
+
   if obj.status.conditions ~= nil then
     for i, condition in ipairs(obj.status.conditions) do
       if condition.type == "ChildResourcesHealthy" then
         healthyCondition = condition
       end
     end
   end
-
-  if obj.metadata.generation == obj.status.observedGeneration then
-    if (healthyCondition ~= {} and healthyCondition.status == "False" and (obj.metadata.generation == healthyCondition.observedGeneration) and healthyCondition.reason == "ISBSvcFailed") or obj.status.phase == "Failed" then
-      hs.status = "Degraded"
-      if obj.status.phase == "Failed" then
-        hs.message = obj.status.message
-      else
-        hs.message = healthyCondition.message
-      end
-      return hs
-    elseif healthyCondition ~= {} and healthyCondition.status == "True" and (obj.metadata.generation == healthyCondition.observedGeneration) and obj.status.phase == "Deployed" then
-      hs.status = "Healthy"
+  
+  if (healthyCondition ~= {} and healthyCondition.status == "False" and healthyCondition.reason == "ISBSvcFailed") or obj.status.phase == "Failed" then
+    hs.status = "Degraded"
+    if obj.status.phase == "Failed" then
+      hs.message = obj.status.message
+    else
       hs.message = healthyCondition.message
-      return hs
     end
+    return hs
+  elseif healthyCondition ~= {} and healthyCondition.status == "False" and healthyCondition.reason == "Progressing" then
+    hs.status = "Progressing"
+    hs.message = healthyCondition.message
+    return hs
+  elseif healthyCondition ~= {} and healthyCondition.status == "True" and obj.status.phase == "Deployed" then
+    hs.status = "Healthy"
+    hs.message = healthyCondition.message
+    return hs
   end
+
+else -- if there's no Status at all, we haven't been reconciled
+  hs.status = "Progressing"
+  hs.message = "Not yet reconciled"
+  return hs
```

Comment:
Could this check be moved to the top so that we can un-nest the non-nil case?

Reply:
Hey Michael. Not sure if I understand - you're referring to this block?:
```
 hs.status = "Progressing"
  hs.message = "Not yet reconciled"
  return hs
```

How is it being nested? I believe it's like this, right?:
```
  if obj.status ~= nil then
    ...
  else -- if there's no Status at all, we haven't been reconciled
      hs.status = "Progressing"
      hs.message = "Not yet reconciled"
      return hs
  end
```

Reply:
I think it makes sense to nest the `Not yet reconciled` case. But I think we should do that earlier in the code and `return` if we hit that case.

By checking early and returning early, we allow the big block currently under `obj.status ~= nil` to be un-nested.

Added a partial suggestion here to help clarify: https://github.com/argoproj/argo-cd/pull/20544/files#r1821127084

Reply:
ahh yeah, that seems good